uat/custom-components/client-java-paho/src/main/java/com/aws/greengrass/testing/util/SslUtil.java

```diff
@@ -0,0 +1,109 @@
+import java.io.FileReader;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.security.GeneralSecurityException;
+import java.security.KeyPair;
+import java.security.KeyStore;
+import java.security.Security;
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManagerFactory;
+
+public final class SslUtil {
+    private SslUtil() {
+    }
+
+    /**
+     * generate SSL socket factory.
+     *
+     * @param connectionParams MQTT connection parameters
+     * @throws IOException on errors
+     * @throws GeneralSecurityException on errors
+     */
+    public static SSLSocketFactory getSocketFactory(MqttLib.ConnectionParams connectionParams)
+            throws IOException, GeneralSecurityException {
+        return getSocketFactory(connectionParams.getCa(), connectionParams.getCert(), connectionParams.getKey());
+    }
+
+    /**
+     * generate SSL socket factory.
+     *
+     * @param caCrtFile client certification
+     * @param crtFile client certification
+     * @param keyFile client certification
+     * @throws IOException on errors
+     * @throws GeneralSecurityException on errors
+     */
+    @SuppressWarnings({"PMD.CloseResource", "PMD.AvoidFileStream"})
+    public static SSLSocketFactory getSocketFactory(final String caCrtFile, final String crtFile, final String keyFile)
+            throws IOException, GeneralSecurityException {
+        Security.addProvider(new BouncyCastleProvider());
+
+        // load CA certificate
+        X509Certificate caCert = null;
+
+        BufferedInputStream bis = new BufferedInputStream(Files.newInputStream(Paths.get(caCrtFile)));
+        CertificateFactory cf = CertificateFactory.getInstance("X.509");
+
+        while (bis.available() > 0) {
+            caCert = (X509Certificate) cf.generateCertificate(bis);
+        }
+
+        // load client certificate
+        bis = new BufferedInputStream(Files.newInputStream(Paths.get(crtFile)));
+        X509Certificate cert = null;
+        while (bis.available() > 0) {
+            cert = (X509Certificate) cf.generateCertificate(bis);
+        }
+
+        // CA certificate is used to authenticate server
+        KeyStore caKs = KeyStore.getInstance(KeyStore.getDefaultType());
+        caKs.load(null, null);
+        caKs.setCertificateEntry("ca-certificate", caCert);
+        TrustManagerFactory tmf = TrustManagerFactory.getInstance("X509");
+        tmf.init(caKs);
+
+        // client key and certificates are sent to server so it can authenticate
+        KeyStore ks = KeyStore.getInstance(KeyStore.getDefaultType());
+        ks.load(null, null);
+        ks.setCertificateEntry("certificate", cert);
+
+        // load client private key
+        PEMParser pemParser = new PEMParser(new FileReader(keyFile));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *pemParser*. The resource is referenced at lines 90, 91. The resource is closed at line 93. There are other execution paths that do not contain closure statements, for example, when PEMParser.readObject throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).

Reply:
fixed